File: helix-runtime/apps/android/app/src/main/java/com/helix/tasks/TasksModels.kt
package com.helix.tasks

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// MARK: - Task Models

@Serializable
data class Task(
    val id: String,
    @SerialName("user_id")
    val userId: String,
    @SerialName("board_id")
    val boardId: String,
    val title: String,
    val description: String? = null,
    val status: TaskStatus = TaskStatus.Todo,
    val priority: TaskPriority = TaskPriority.Medium,
    @SerialName("due_date")
    val dueDate: String? = null,
    @SerialName("completed_at")
    val completedAt: String? = null,
    @SerialName("assignee_id")
    val assigneeId: String? = null,
    @SerialName("assignee_name")
    val assigneeName: String? = null,
    @SerialName("assignee_email")
    val assigneeEmail: String? = null,
    @SerialName("estimated_hours")
    val estimatedHours: Double? = null,
    @SerialName("actual_hours")
    val actualHours: Double? = null,
    @SerialName("subtask_count")
    val subtaskCount: Int = 0,
    @SerialName("completed_subtask_count")
    val completedSubtaskCount: Int = 0,
    val tags: List<String> = emptyList(),
    @SerialName("depends_on_task_ids")
    val dependsOnTaskIds: List<String> = emptyList(),
    @SerialName("blocks_task_ids")
    val blocksTaskIds: List<String> = emptyList(),
    @SerialName("created_at")
    val createdAt: String,
    @SerialName("updated_at")
    val updatedAt: String,
    @SerialName("deleted_at")
    val deletedAt: String? = null,
    @SerialName("is_deleted")
    val isDeleted: Boolean = false,
) {
    val isCompleted: Boolean
        get() = status == TaskStatus.Done

    val progressPercentage: Double
        get() = if (subtaskCount > 0) completedSubtaskCount.toDouble() / subtaskCount else if (isCompleted) 1.0 else 0.0

    val isOverdue: Boolean
        get() = dueDate != null && !isCompleted && dueDate < System.currentTimeMillis().toString()
}

@Serializable
enum class TaskStatus {
    @SerialName("todo")
    Todo,

    @SerialName("in_progress")
    InProgress,

    @SerialName("in_review")
    InReview,

    @SerialName("done")
    Done,
}

@Serializable
enum class TaskPriority {
    @SerialName("low")
    Low,

    @SerialName("medium")
    Medium,

    @SerialName("high")
    High,

    @SerialName("critical")
    Critical,
}

@Serializable
data class Subtask(
    val id: String,
    @SerialName("task_id")
    val taskId: String,
    val title: String,
    @SerialName("is_completed")
    val isCompleted: Boolean = false,
    @SerialName("completed_at")
    val completedAt: String? = null,
    @SerialName("created_at")
    val createdAt: String,
    @SerialName("updated_at")
    val updatedAt: String,
)

// MARK: - Kanban Board Models

@Serializable
data class KanbanBoard(
    val id: String,
    @SerialName("user_id")
    val userId: String,
    val name: String,
    val description: String? = null,
    @SerialName("is_default")
    val isDefault: Boolean = false,
    @SerialName("column_order")
    val columnOrder: List<String> = emptyList(),
    @SerialName("created_at")
    val createdAt: String,
    @SerialName("updated_at")
    val updatedAt: String,
)

// MARK: - Focus Time Models

@Serializable
data class FocusTime(
    @SerialName("daily_goal_minutes")
    val dailyGoalMinutes: Int = 60,
    @SerialName("start_hour")
    val startHour: Int = 9,
    @SerialName("end_hour")
    val endHour: Int = 17,
    @SerialName("enable_notifications")
    val enableNotifications: Boolean = true,
    @SerialName("blocked_apps_url_patterns")
    val blockedAppsUrlPatterns: List<String> = emptyList(),
)

// MARK: - Time Tracking Models

@Serializable
data class TimeEntry(
    val id: String,
    @SerialName("task_id")
    val taskId: String,
    @SerialName("user_id")
    val userId: String,
    val description: String? = null,
    @SerialName("duration_minutes")
    val durationMinutes: Int,
    @SerialName("started_at")
    val startedAt: String,
    @SerialName("ended_at")
    val endedAt: String,
)

// MARK: - Analytics Models

@Serializable
data class TaskAnalytics(
    @SerialName("total_tasks")
    val totalTasks: Int,
    @SerialName("completed_tasks")
    val completedTasks: Int,
    @SerialName("in_progress_tasks")
    val inProgressTasks: Int,
    @SerialName("overdue_tasks")
    val overdueTasks: Int,
    @SerialName("average_completion_days")
    val averageCompletionDays: Double,
    @SerialName("completion_rate")
    val completionRate: Double,
    @SerialName("velocity_this_week")
    val velocityThisWeek: Int,
    @SerialName("velocity_last_week")
    val velocityLastWeek: Int,
    @SerialName("priority_distribution")
    val priorityDistribution: Map<String, Int> = emptyMap(),
    @SerialName("status_distribution")
    val statusDistribution: Map<String, Int> = emptyMap(),
)

// MARK: - Search Filter Models

@Serializable
data class TaskSearchFilter(
    val query: String = "",
    val status: List<TaskStatus> = emptyList(),
    val priority: List<TaskPriority> = emptyList(),
    @SerialName("assignee_id")
    val assigneeId: String? = null,
    @SerialName("board_id")
    val boardId: String? = null,
    val tags: List<String> = emptyList(),
    @SerialName("has_subtasks")
    val hasSubtasks: Boolean? = null,
    @SerialName("is_overdue")
    val isOverdue: Boolean? = null,
)

// MARK: - Error Handling

sealed class TaskError : Exception() {
    object ServiceUnavailable : TaskError()
    object InvalidTaskData : TaskError()
    object TaskNotFound : TaskError()
    object UnauthorizedAccess : TaskError()
    data class NetworkError(val message: String) : TaskError()
    data class DecodingError(val message: String) : TaskError()
    object UnknownError : TaskError()

    val localizedMessage: String
        get() = when (this) {
            ServiceUnavailable -> "Task service is unavailable. Please try again later."
            InvalidTaskData -> "Invalid task data provided."
            TaskNotFound -> "Task not found."
            UnauthorizedAccess -> "You don't have permission to access this task."
            is NetworkError -> "Network error: ${this.message}"
            is DecodingError -> "Failed to decode task data: ${this.message}"
            UnknownError -> "An unknown error occurred."
        }
}

File: helix-runtime/apps/ios/Sources/Tasks/TasksModels.swift
import Foundation

// MARK: - Task Models

struct Task: Identifiable, Codable, Equatable, Hashable {
    let id: String
    let userId: String
    let boardId: String

    var title: String
    var description: String
    var status: TaskStatus
    var priority: TaskPriority
    var dueDate: Date?
    var completedAt: Date?

    var assigneeId: String?
    var assigneeName: String?
    var assigneeEmail: String?

    var estimatedHours: Double?
    var actualHours: Double?
    var subtaskCount: Int
    var completedSubtaskCount: Int

    var tags: [String]
    var dependsOnTaskIds: [String]
    var blocksTaskIds: [String]

    var createdAt: Date
    var updatedAt: Date
    var deletedAt: Date?

    var isDeleted: Bool

    enum CodingKeys: String, CodingKey {
        case id, userId, boardId, title, description, status, priority
        case dueDate = "due_date"
        case completedAt = "completed_at"
        case assigneeId = "assignee_id"
        case assigneeName = "assignee_name"
        case assigneeEmail = "assignee_email"
        case estimatedHours = "estimated_hours"
        case actualHours = "actual_hours"
        case subtaskCount = "subtask_count"
        case completedSubtaskCount = "completed_subtask_count"
        case tags
        case dependsOnTaskIds = "depends_on_task_ids"
        case blocksTaskIds = "blocks_task_ids"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case deletedAt = "deleted_at"
        case isDeleted = "is_deleted"
    }

    var isCompleted: Bool {
        status == .done
    }

    var progressPercentage: Double {
        guard subtaskCount > 0 else { return isCompleted ? 1.0 : 0.0 }
        return Double(completedSubtaskCount) / Double(subtaskCount)
    }

    var isOverdue: Bool {
        guard let dueDate = dueDate, !isCompleted else { return false }
        return dueDate < Date()
    }
}

enum TaskStatus: String, Codable {
    case todo
    case inProgress = "in_progress"
    case inReview = "in_review"
    case done

    var displayName: String {
        switch self {
        case .todo:
            return "To Do"
        case .inProgress:
            return "In Progress"
        case .inReview:
            return "In Review"
        case .done:
            return "Done"
        }
    }
}

enum TaskPriority: String, Codable {
    case low
    case medium
    case high
    case critical

    var displayName: String {
        rawValue.capitalized
    }
}

struct Subtask: Identifiable, Codable, Equatable, Hashable {
    let id: String
    let taskId: String

    var title: String
    var isCompleted: Bool
    var completedAt: Date?

    var createdAt: Date
    var updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, taskId, title
        case isCompleted = "is_completed"
        case completedAt = "completed_at"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

// MARK: - Kanban Board Models

struct KanbanBoard: Identifiable, Codable, Equatable, Hashable {
    let id: String
    let userId: String

    var name: String
    var description: String?
    var isDefault: Bool
    var columnOrder: [String] // column IDs in order

    var createdAt: Date
    var updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, userId, name, description
        case isDefault = "is_default"
        case columnOrder = "column_order"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

struct KanbanColumn: Identifiable, Codable, Equatable, Hashable {
    let id: String
    let boardId: String

    var name: String
    var status: TaskStatus
    var position: Int
    var wipLimit: Int? // Work in progress limit

    var createdAt: Date
    var updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, boardId, name, status, position
        case wipLimit = "wip_limit"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

// MARK: - Focus Time Models

struct FocusSession: Identifiable, Codable, Equatable, Hashable {
    let id: String
    let userId: String
    let taskId: String

    var startTime: Date
    var endTime: Date?
    var isCompleted: Bool

    var durationMinutes: Int {
        let end = endTime ?? Date()
        return Int(end.timeIntervalSince(startTime) / 60)
    }
}

struct FocusTime: Codable, Equatable, Hashable {
    var dailyGoalMinutes: Int
    var startHour: Int // 0-23
    var endHour: Int // 0-23
    var enableNotifications: Bool
    var blockedAppsUrlPatterns: [String]

    enum CodingKeys: String, CodingKey {
        case dailyGoalMinutes = "daily_goal_minutes"
        case startHour = "start_hour"
        case endHour = "end_hour"
        case enableNotifications = "enable_notifications"
        case blockedAppsUrlPatterns = "blocked_apps_url_patterns"
    }
}

// MARK: - Time Tracking Models

struct TimeEntry: Identifiable, Codable, Equatable, Hashable {
    let id: String
    let taskId: String
    let userId: String

    var description: String?
    var durationMinutes: Int
    var startedAt: Date
    var endedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, taskId, userId, description
        case durationMinutes = "duration_minutes"
        case startedAt = "started_at"
        case endedAt = "ended_at"
    }
}

struct TimeTrackingStats: Codable, Equatable, Hashable {
    var totalMinutesTracked: Int
    var totalMinutesEstimated: Int
    var tasksTracked: Int
    var averageAccuracy: Double // actual vs estimated
    var dailyTotals: [String: Int] // ISO8601 date -> minutes

    enum CodingKeys: String, CodingKey {
        case totalMinutesTracked = "total_minutes_tracked"
        case totalMinutesEstimated = "total_minutes_estimated"
        case tasksTracked = "tasks_tracked"
        case averageAccuracy = "average_accuracy"
        case dailyTotals = "daily_totals"
    }
}

// MARK: - Analytics Models

struct TaskAnalytics: Codable, Equatable, Hashable {
    var totalTasks: Int
    var completedTasks: Int
    var inProgressTasks: Int
    var overdueTasks: Int

    var averageCompletionDays: Double
    var completionRate: Double
    var velocityThisWeek: Int // tasks completed this week
    var velocityLastWeek: Int

    var priorityDistribution: [String: Int]
    var statusDistribution: [String: Int]

    enum CodingKeys: String, CodingKey {
        case totalTasks = "total_tasks"
        case completedTasks = "completed_tasks"
        case inProgressTasks = "in_progress_tasks"
        case overdueTasks = "overdue_tasks"
        case averageCompletionDays = "average_completion_days"
        case completionRate = "completion_rate"
        case velocityThisWeek = "velocity_this_week"
        case velocityLastWeek = "velocity_last_week"
        case priorityDistribution = "priority_distribution"
        case statusDistribution = "status_distribution"
    }
}

// MARK: - Search & Filter Models

struct TaskSearchFilter: Codable, Equatable, Hashable {
    var query: String
    var status: [TaskStatus]
    var priority: [TaskPriority]
    var assigneeId: String?
    var boardId: String?
    var tags: [String]
    var dueDateRange: DateRange?
    var hasSubtasks: Bool?
    var isOverdue: Bool?

    enum CodingKeys: String, CodingKey {
        case query, status, priority
        case assigneeId = "assignee_id"
        case boardId = "board_id"
        case tags
        case dueDateRange = "due_date_range"
        case hasSubtasks = "has_subtasks"
        case isOverdue = "is_overdue"
    }
}

struct DateRange: Codable, Equatable, Hashable {
    var startDate: Date
    var endDate: Date

    enum CodingKeys: String, CodingKey {
        case startDate = "start_date"
        case endDate = "end_date"
    }
}

// MARK: - Error Handling

enum TaskError: LocalizedError {
    case serviceUnavailable
    case invalidTaskData
    case taskNotFound
    case unauthorizedAccess
    case networkError(String)
    case decodingError(String)
    case unknownError

    var errorDescription: String? {
        switch self {
        case .serviceUnavailable:
            return "Task service is unavailable. Please try again later."
        case .invalidTaskData:
            return "Invalid task data provided."
        case .taskNotFound:
            return "Task not found."
        case .unauthorizedAccess:
            return "You don't have permission to access this task."
        case .networkError(let message):
            return "Network error: \(message)"
        case .decodingError(let message):
            return "Failed to decode task data: \(message)"
        case .unknownError:
            return "An unknown error occurred."
        }
    }
}
